librz/util/float/float.c

```diff
@@ -0,0 +1,1884 @@
+// SPDX-FileCopyrightText: 2022 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/**
+ * \file float.c
+ * This file implements IEEE-754 binary float number operations (32/64/128)
+ * IEEE binary representations, use binary digits to represent float. machine-friendly
+ * binary32 format (single) : use a 32 bits bitvector to represent float
+ * 	32 bits = 1 (sign bit) + 8 (exponent bits) + 23 (mantissa bits)
+ * 	exponent value range : -126 ~ 127
+ * binary64 format (double) : use a 64 bits bitvector to represent float
+ *  	64 bits = 1 (sign bit) + 11 (exponent bits) + 52 (mantissa bits)
+ * 	exponent value range : -1022 ~ 1023
+ * binary128 format, use a 128 bits bitvector to represent float
+ *  	128 bits = 1 (sign bit) + 15 (exponent bits) + 112 (mantissa bits)
+ * 	exponent value range : -16382 ~ 16383
+ **/
+
+#include "float_internal.c"
+
+/**
+ * \brief return the bitvector string of a float
+ * \param f float
+ * \return char* string of bitvector
+ */
+RZ_API RZ_OWN char *rz_float_as_bit_string(RZ_NULLABLE RzFloat *f) {
+	if (!f || !f->s) {
+		return NULL;
+	}
+	return rz_bv_as_string(f->s);
+}
+
+/**
+ * \brief return the bitvector hex string of a float
+ * \param f float
+ * \param use_pad use padding before the hex string
+ * \return char* hex string of bitvector
+ */
+RZ_API RZ_OWN char *rz_float_as_hex_string(RZ_NULLABLE RzFloat *f, bool use_pad) {
+	if (!f || !f->s) {
+		return NULL;
+	}
+	return rz_bv_as_hex_string(f->s, use_pad);
+}
+
+/**
+ * \brief return a human-readable string of float
+ * \param f float
+ * \return a human-readable string of float.
+ * exponent part and mantissa part would be split as follows:
+ * 'sign' 'exponent part' | 'mantissa part'
+ *  1.0f would be shown as +01111111|00000000000000000000000
+ */
+RZ_API RZ_OWN char *rz_float_as_string(RZ_NONNULL RzFloat *f) {
+	rz_return_val_if_fail(f && f->s, NULL);
+
+	ut32 man_len = rz_float_get_format_info(f->r, RZ_FLOAT_INFO_MAN_LEN);
+	ut32 exp_len = rz_float_get_format_info(f->r, RZ_FLOAT_INFO_EXP_LEN);
+	ut32 total = rz_float_get_format_info(f->r, RZ_FLOAT_INFO_TOTAL_LEN);
+
+	char *str = (char *)malloc(total + 2);
+	if (!str) {
+		return NULL;
+	}
+
+	ut32 pos = rz_bv_len(f->s) - 1;
+	ut32 i;
+
+	str[0] = rz_float_get_sign(f) ? '-' : '+';
```

Comment:
shall we always put + on positive numbers? (do not change this for now)

Reply:
Not sure about this.

Reply:
not necessary for positive numbers.